[Forum] Display who upvoted/downvoted each post. Closes #346

stopnoanime opened the pull request:
Displays usernames of users who upvoted/downvoted each post in a tooltip.
Only 10 usernames for each post are queried and displayed to improve performance.
I'm using prefetch_related to query necessary reactions for all posts with a single query.

### To Do:
- Add tests
- Improve translation ( I don't know if I should also change files in /locale )

twalen commented on oioioi/forum/templatetags/display_reacted_by.py:
```diff
@@ -0,0 +1,23 @@
+
+@register.simple_tag
+def display_reacted_by(post, rtype):
+  # post needs to be queried with prefetch_reacted_by
```
I suggest to add sanity check if rtype is "upvote" or "downvote" and tigger an error if incorrect rtype is used.

stopnoanime replied:
Added the check.

twalen commented on oioioi/forum/templatetags/display_reacted_by.py:
```diff
@@ -0,0 +1,23 @@
+
+  output = ', '.join([
+    get_user_display_name(reaction.author)
+    for reaction in getattr(post, rtype.lower() + 'd_by')
```
why `.lower()`, since it is an internal tag we can expect that someone will use it with proper rtype

stopnoanime commented:
This PR has a bug that causes the latest posts view to not load. I turned the PR to a draft while I investigate why the tests didn't detect this.

stopnoanime commented:
Added a test that checks the latest post view with reactions enabled.
Added two dictionaries that map reactions to necessary attribute names so the code doesn't rely on string manipulations anymore. Also moved prefetching reacted_by into the Post model so there is no need to call a separate function to do so.

twalen commented on oioioi/forum/models.py:
```diff
@@ -192,22 +193,26 @@ class Post(models.Model):
+                    models.Prefetch(
+                        'reactions', 
+                        to_attr=attr_name,
+                        queryset=PostReaction.objects.filter(type_of_reaction=rtype).select_related('author')[:max_count], 
```
it would be nice to add some ordering here, otherwise we can get random results (especially important during tests): i.e. `order_by('-pk')`

also you can add Meta class to PostReaction like:
```
class Meta:
        ordering = ['-pk']
```

stopnoanime replied:
Done.

twalen reviewed (changes requested):
Looks ok, but please add ordering + modify tests

stopnoanime commented:
Tests started failing due to changing Meta options on PostReaction, will need to figure it out.

stopnoanime commented:
Adding the migration fixed the issue. I didn't realize a change like this would require a migration.

stopnoanime commented:
Decided to undo the meta ordering change and migration. The meta ordering seems unnecessary and adding it requires a migration file.
